qiskit/optimization/algorithms/optimization_algorithm.py

```diff
@@ -104,24 +104,54 @@ class OptimizationResult:
         results: The original results object returned from the optimization algorithm. This can
             contain more information than only the optimal value and function value.
         status: The termination status of the algorithm.
+        variables: The list of variables under optimization.
     """
 
     Status = OptimizationResultStatus
 
     def __init__(self, x: Optional[Any] = None, fval: Optional[Any] = None,
                  results: Optional[Any] = None,
-                 status: OptimizationResultStatus = OptimizationResultStatus.SUCCESS) -> None:
-        self._val = x
+                 status: OptimizationResultStatus = OptimizationResultStatus.SUCCESS,
+                 variables: Optional[List[Variable]] = None) -> None:
+        self._x = x  # pylint: disable=invalid-name
+        self._variables = variables
         self._fval = fval
         self._results = results
         self._status = status
 
     def __repr__(self):
-        return '([{}] / {} / {})'.format(','.join([str(x_) for x_ in self.x]), self.fval,
-                                         self.status)
+        return 'optimal variables: [{}]\noptimal function value: {}\nstatus: {}' \
+            .format(','.join([str(x_) for x_ in self.x]), self.fval, self.status.name)
```

Comment:
I think self.x needs to be ensured that it is not None before its used in the comprehension otherwise it will fail with a non-iterable error. I realize that the constructor change means that None passed through there will end up in an empty list but the setter seems to still allow x to be set directly to None.

Reply:
I think that following line ensures it is not None
116: self._x = x if x else []

Reply:
The original version has the same issue, i.e., `[str(x_) for x_ in self.x]`.
I think allowing `self.x = None` makes all APIs implementation tricky. So, I would like to investigate whether we can remove `Optional[Any]` stuff in the constructor.

Reply:
in __init__ method, if self.x is None, it's set to [ ].
However, python lacks the feature that would protect class variables from being set from outside.

Reply:
@t-imamichi The setter allows None to pass in too. If its left as Any in constructor None will should still o be handled there - it just would not default to None as it currently does. Is there not a more specific list of type(s) this can can be restricted to - Any seems so wide open but maybe thats required?

Reply:
>However, python lacks the feature that would protect class variables from being set from outside.

For sure someone can go in and set a variable we have annotated as private i.e. _x to None. All bets are off for that. Here its more managing things via the public API - i.e. constructor and setter, so there is some expected behavior in the presence of what is a valid value of None being passed through the public API.